### ISSUE TITLE
Profiling tool

### ISSUE THREAD

**Issue by CanyonWind** [HazyResearch/flash-attention#38]
Hi, thank you for the great work. Just wonder what tool did you use to do the profiling and successfully identified that IO is the bottleneck instead of the actual computation -- is the tool based on pytorch/cuda? Thanks ahead for any pointers.

**Comment by tridao** [comment]
We were using the [Pytorch](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html) [profiler](https://pytorch.org/tutorials/intermediate/tensorboard_profiler_tutorial.html), which would tell us that masking, softmax, and dropout take more time than matmul in the attention computation. 

The Nvidia deep learning [performance](https://docs.nvidia.com/deeplearning/performance/index.html) [guide](https://docs.nvidia.com/deeplearning/performance/dl-performance-gpu-background/index.html) is also very helpful.

Nsight [Compute](https://developer.nvidia.com/nsight-compute) is a profiler from Nvidia that gives more information about each kernel (how many memory reads/writes, how many FLOPs, is it maximally using the resources).

**Comment by CanyonWind** [comment]
Thanks for the fast response, this helps a lot!
